frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -928,7 +914,53 @@ pub mod pallet {
 	pub struct Pallet<T>(PhantomData<T>);
 }
 
+impl<T: Config> sp_runtime::offchain::storage_lock::BlockNumberProvider for Pallet<T> {
+		type BlockNumber = T::BlockNumber;
+		fn current_block_number() -> Self::BlockNumber {
+			<frame_system::Pallet<T>>::block_number()
+		}
+	}
+
 impl<T: Config> Pallet<T> {
+	/// Internal logic of the offchain worker, to be executed only when the offchain lock is
+	/// acquired with success.
+	fn do_synchronized_offchain_worker(now: T::BlockNumber) {
+		log!(trace, "lock for offchain worker acquired.");
+		match Self::current_phase() {
+			Phase::Unsigned((true, opened)) if opened == now => {
+				// mine a new solution, cache it, and attempt to submit it
+				let initial_output = Self::ensure_offchain_repeat_frequency(now).and_then(|_| {
+					Self::mine_check_save_submit()
+				});
+				log!(debug, "initial offchain thread output: {:?}", initial_output);
+			}
+			Phase::Unsigned((true, opened)) if opened < now => {
+				// try and resubmit the cached solution, and recompute ONLY if it is not
+				// feasible.
+				let resubmit_output = Self::ensure_offchain_repeat_frequency(now).and_then(|_| {
+					Self::restore_or_compute_then_maybe_submit()
+				});
+				log!(debug, "resubmit offchain thread output: {:?}", resubmit_output);
+			}
+			_ => {}
+		}
+
+		// after election finalization, clear OCW solution storage.
+		if <frame_system::Pallet<T>>::events()
+			.into_iter()
+			.filter_map(|event_record| {
+				let local_event = <T as Config>::Event::from(event_record.event);
+				local_event.try_into().ok()
+			})
+			.find(|event| {
+				matches!(event, Event::ElectionFinalized(_))
+			})
+			.is_some()
+		{
+			unsigned::kill_ocw_solution::<T>();
+		}
```

Comment:
What if this function is not run on this specific block due to the newly introduced lock ?

It shouldn't harm too much, as it will be removed due to failing feasibility.

Reply:
do you suggest anywhere else where we would kill the solution?

Thinking about it again, if we simply don't kill this, it will be replaced with a fresh one on the next round. In the past it would have been tricky since we could race around it. But now, the initial OCW will attempt to mine a new solution and replace this, and no other OCW will work in parallel, so it should be safe.

Reply:
I don't have very convincing suggestion, maybe we could match the phase and remove when the phase is close, but removing on every block when the phase is close is a bit useless too.

anyway I agree with what you say, current implementation is fine, even if not removed, it is not an issue